Enable setting the hostName as the ipAddress

Spring Cloud Eureka client configuration allows using the IP Address of the host to be used as the host name rather than the hostname that is by default assigned. This is particularly helpful in AWS environments where the hostname is not in many cases resolvable by your DNS server. E.g. the default host name that AWS assigns is of the type ip-XX-XX-XX-XX and this many times does not resolved depending on how your DNS is configured

Spring Cloud supplies a config parameter preferIpAddress which is by default false and when set to true will use the local-ipv4 / public-ipv4 instead of local-hostname / public-hostname attributes from the EC2 metadata. 

Can we add this feature since without this we are unable to use the module in our environment. I will be happy to work on this and issue a pull request

Hi @jquatier Thanks for a great library. Currently we are having Spring and NodeJS base micro services app which has a Spring API Gateway (written in Java). The Gateway can access the Java base micro services through Feign client without any problem.

But when trying to access the NodeJS based micro service, we always get "Caused by: com.netflix.client.ClientException: Load balancer does not have available server for client: MEMBER-SERVICE"

The Node-Eureka client is 4.1.4 and has the following settings:
const EUREKA_CLIENT = new EUREKA({
  instance: {
    app: config.application_identifier,
    //hostName: config.server_name, 
    //ipAddr: '127.0.0.1',
    vipAddress: 'test.something.com',
    statusPageUrl: 'http://__HOST__:30043/',
    healthCheckUrl: 'http://__HOST__:30043/healthcheck',
    preferIpAddress : true,
    port: {
      '$': port,
      '@enabled': 'true',
    },
    secureVipAddress: config.application_identifier,
    /*dataCenterInfo: {
      '@class': 'com.netflix.appinfo.InstanceInfo$DefaultDataCenterInfo',
      name: 'MyOwn',
    }*/
     dataCenterInfo: {
      '@class': 'com.netflix.appinfo.AmazonInfo',
      name: 'Amazon',
    }
  },

But as you can see based on the image of eureka, it's still having the local host name as apposed to the local IP. The concern is, since preferIpAddress is not getting set, the local host name is used, thus the server can't be resolved?

![issue](https://cloud.githubusercontent.com/assets/20202478/22119220/b9a6ab1e-dea0-11e6-8e30-5f989956094c.png)

Appreciate if you can give us an input on this. Again, thanks a lot for the hard work.

--- a/README.md
+++ b/README.md
@@ -132,6 +132,7 @@ const client = new Eureka({
 
 Notes:
   - Under this configuration, the instance `hostName` and `ipAddr` will be set to the public host and public IP that the AWS metadata provides. You can set `eureka.useLocalMetadata` to `true` to use the private host and private IP address instead.
+  - If you want to register using the IP address as the hostname, set `eureka.preferIpAddress` to `true`. This may be used in combination with `eureka.useLocalMetadata` for selecting the private or public IP.
   - For status and healthcheck URLs, you may use the replacement key of `__HOST__` to use the host from the metadata.
   - Metadata fetching can be disabled by setting `config.eureka.fetchMetadata` to `false` if you want to provide your own metadata in AWS environments.
 
@@ -205,6 +206,7 @@ option | default value | description
 `eureka.clusterRefreshInterval` | `300000` | milliseconds to wait between refreshing cluster hosts (DNS resolution only)
 `eureka.fetchMetadata` | `true` | fetch AWS metadata when in AWS environment, see [Configuring for AWS environments](#configuring-for-aws-environments)
 `eureka.useLocalMetadata` | `false` | use local IP and local hostname from metadata when in an AWS environment.
+`eureka.preferIpAddress` | `false` | use IP address (local or public) as the hostname for registration when in an AWS environment.
 
 ## Events
 

--- a/src/EurekaClient.js
+++ b/src/EurekaClient.js
@@ -412,7 +412,8 @@ export default class Eureka extends EventEmitter {
     Fetches the metadata using the built-in client and updates the instance
     configuration with the hostname and IP address. If the value of the config
     option 'eureka.useLocalMetadata' is true, then the local IP address and
-    hostname is used. Otherwise, the public IP address and hostname is used.
+    hostname is used. Otherwise, the public IP address and hostname is used. If
+    'eureka.preferIpAddress' is true, the IP address will be used as the hostname.
 
     A string replacement is done on the healthCheckUrl, statusPageUrl and
     homePageUrl so that users can define the URLs with a placeholder for the
@@ -431,23 +432,25 @@ export default class Eureka extends EventEmitter {
         metadataResult
       );
       const useLocal = this.config.eureka.useLocalMetadata;
+      const preferIpAddress = this.config.eureka.preferIpAddress;
       const metadataHostName = metadataResult[useLocal ? 'local-hostname' : 'public-hostname'];
-      this.config.instance.hostName = metadataHostName;
-      this.config.instance.ipAddr = metadataResult[useLocal ? 'local-ipv4' : 'public-ipv4'];
+      const metadataIpAddress = metadataResult[useLocal ? 'local-ipv4' : 'public-ipv4'];
+      this.config.instance.hostName = preferIpAddress ? metadataIpAddress : metadataHostName;
+      this.config.instance.ipAddr = metadataIpAddress;
 
       if (this.config.instance.statusPageUrl) {
         const { statusPageUrl } = this.config.instance;
-        const replacedUrl = statusPageUrl.replace('__HOST__', metadataHostName);
+        const replacedUrl = statusPageUrl.replace('__HOST__', this.config.instance.hostName);
         this.config.instance.statusPageUrl = replacedUrl;
       }
       if (this.config.instance.healthCheckUrl) {
         const { healthCheckUrl } = this.config.instance;
-        const replacedUrl = healthCheckUrl.replace('__HOST__', metadataHostName);
+        const replacedUrl = healthCheckUrl.replace('__HOST__', this.config.instance.hostName);
         this.config.instance.healthCheckUrl = replacedUrl;
       }
       if (this.config.instance.homePageUrl) {
         const { homePageUrl } = this.config.instance;
-        const replacedUrl = homePageUrl.replace('__HOST__', metadataHostName);
+        const replacedUrl = homePageUrl.replace('__HOST__', this.config.instance.hostName);
         this.config.instance.homePageUrl = replacedUrl;
       }
 

--- a/src/defaultConfig.js
+++ b/src/defaultConfig.js
@@ -15,6 +15,7 @@ export default {
     clusterRefreshInterval: 300000,
     fetchMetadata: true,
     useLocalMetadata: false,
+    preferIpAddress: false,
   },
   instance: {},
 };

--- a/test/EurekaClient.test.js
+++ b/test/EurekaClient.test.js
@@ -867,6 +867,26 @@ describe('Eureka client', () => {
       expect(client.config.instance.homePageUrl).to.equal('http://ec2-127-0-0-1.us-fake-1.mydomain.com:8080/');
     });
 
+    it('should update hosts with AWS metadata public IP when preferIpAddress === true', () => {
+      // Setup
+      config = {
+        instance: instanceConfig,
+        eureka: { host: '127.0.0.1', port: 9999, preferIpAddress: true },
+      };
+      client = new Eureka(config);
+      metadataSpy = sinon.spy();
+
+      sinon.stub(client.metadataClient, 'fetchMetadata').yields(awsMetadata);
+
+      // Act
+      client.addInstanceMetadata(metadataSpy);
+      expect(client.config.instance.hostName).to.equal('54.54.54.54');
+      expect(client.config.instance.ipAddr).to.equal('54.54.54.54');
+      expect(client.config.instance.statusPageUrl).to.equal('http://54.54.54.54:8080/info');
+      expect(client.config.instance.healthCheckUrl).to.equal('http://54.54.54.54:8077/healthcheck');
+      expect(client.config.instance.homePageUrl).to.equal('http://54.54.54.54:8080/');
+    });
+
     it('should update hosts with AWS metadata local host if useLocalMetadata === true', () => {
       // Setup
       config = {
@@ -885,6 +905,27 @@ describe('Eureka client', () => {
       expect(client.config.instance.statusPageUrl).to.equal('http://fake-1:8080/info');
       expect(client.config.instance.healthCheckUrl).to.equal('http://fake-1:8077/healthcheck');
       expect(client.config.instance.homePageUrl).to.equal('http://fake-1:8080/');
+    });
+
+    it('should update hosts with AWS metadata local IP if useLocalMetadata === true' +
+      ' and preferIpAddress === true', () => {
+      // Setup
+      config = {
+        instance: instanceConfig,
+        eureka: { host: '127.0.0.1', port: 9999, useLocalMetadata: true, preferIpAddress: true },
+      };
+      client = new Eureka(config);
+      metadataSpy = sinon.spy();
+
+      sinon.stub(client.metadataClient, 'fetchMetadata').yields(awsMetadata);
+
+      // Act
+      client.addInstanceMetadata(metadataSpy);
+      expect(client.config.instance.hostName).to.equal('10.0.1.1');
+      expect(client.config.instance.ipAddr).to.equal('10.0.1.1');
+      expect(client.config.instance.statusPageUrl).to.equal('http://10.0.1.1:8080/info');
+      expect(client.config.instance.healthCheckUrl).to.equal('http://10.0.1.1:8077/healthcheck');
+      expect(client.config.instance.homePageUrl).to.equal('http://10.0.1.1:8080/');
     });
   });
 

